### PR TITLE
Fixes #555 and #596

### LINKED ISSUE #555: Graciously handle empty paths to by-path
Hi folks!

When the `paths` param to [by-path](https://github.com/boot-clj/boot/blob/2.6.0/boot/core/src/boot/core.clj#L1179) is an empty `seq`, this is the :boom: I receive :

```
                              java.lang.Thread.run              Thread.java:  745
java.util.concurrent.ThreadPoolExecutor$Worker.run  ThreadPoolExecutor.java:  617
 java.util.concurrent.ThreadPoolExecutor.runWorker  ThreadPoolExecutor.java: 1142
               java.util.concurrent.FutureTask.run          FutureTask.java:  266
                             clojure.lang.AFn.call                 AFn.java:   18
               clojure.core/binding-conveyor-fn/fn                 core.clj: 2020
                                 boot.core/boot/fn                 core.clj: 1029
                               boot.core/run-tasks                 core.clj: 1019
              pandeiro.boot-http/eval1138/fn/fn/fn            boot_http.clj:  110
                    boot.task.built-in/fn/fn/fn/fn             built_in.clj:  416
                 boot.task.built-in/fn/fn/fn/fn/fn             built_in.clj:  419
              boot.task.built-in/fn/fn/fn/fn/fn/fn             built_in.clj:  419
                    boot.task.built-in/fn/fn/fn/fn             built_in.clj:  235
               adzerk.boot-reload/eval803/fn/fn/fn          boot_reload.clj:  187
            adzerk.boot-reload/eval803/fn/fn/fn/fn          boot_reload.clj:  188
                        adzerk.boot-reload/changed          boot_reload.clj:   29
                             boot.core/not-by-path                 core.clj: 1266
                        clojure.lang.RestFn.invoke              RestFn.java:  442
                                 boot.core/by-path                 core.clj: 1261
                        clojure.lang.RestFn.invoke              RestFn.java:  442
                          boot.core/file-filter/fn                 core.clj: 1221
                                clojure.core/apply                 core.clj:  657
                       clojure.lang.RestFn.applyTo              RestFn.java:  132
                    clojure.lang.AFn.applyToHelper                 AFn.java:  152
                        clojure.lang.RestFn.invoke              RestFn.java:  399
                       clojure.lang.AFn.throwArity                 AFn.java:  429
```

Potentially this involves also the other `by-*` and `not-by-*` functions. It basically happens because `juxt` does not like to be called in an "empty" apply [this](https://github.com/boot-clj/boot/blob/2.6.0/boot/core/src/boot/core.clj#L1140).

Should be an easy fix in `file-filter` :man_dancing:

### PR DESCRIPTION
**Pull request by hit023**

### CODE REVIEW

**Comment by arichiardi** [comment]
Good job 😀 I left a couple of notes and I would suggest to implement some tests around the new features. Also running the previously existing test is a way to confirm you did not break anything (run `make clean test`).

File: boot/core/src/boot/task/built_in.clj [hunk 1]
```diff
@@ -348,9 +348,10 @@
                        (update-in [:dependencies] #(exclude (filter include? %))))]
           (if-let [conflicts (and safe (not-empty (dep-conflicts env)))]
             (throw (ex-info "Unresolved dependency conflicts." {:conflicts conflicts}))
-            (let [resolved        (pod/resolve-dependency-jars env)
+            (let [source-path     (into [] :source-paths)           
+                  resolved        (pod/resolve-dependency-jars env)
                   relative-paths  (map (partial relativize local-repo) resolved)]
-              (spit file-out (apply str (interpose ":" relative-paths))))))))))
+              (spit file-out (apply str (interpose ":" relative-paths) (interpose ":" source-path))))))))))
```
Comment (Deraen) [inline boot/core/src/boot/task/built_in.clj]: relative paths and source-paths should be separated by ":"

**Comment by arichiardi** [comment]
@hit023 we can refactor this to solve #596 only and keep a separate one for the other, what do you think? Also, can I help with something here?

**Comment by hit023** [comment]
Yes, @arichiardi . I'd love to get on with work, here. But my end-semester exams are underway, so I can resume only afterwards. Thanks.

File: boot/core/src/boot/task/built_in.clj (hunk 1 above)
Comment (arichiardi) [inline boot/core/src/boot/task/built_in.clj]: Is this comment outdated? I see they are, maybe I am misreading?

Comment (arichiardi) [inline boot/core/src/boot/task/built_in.clj]: Oh ok I got what you said

**Comment by arichiardi** [comment]
@hit023 Closing this in favor of #611 and #608

**Pull request closed without merging**